internal/server/server.go

```diff
@@ -106,541 +120,60 @@ func (s *Server) Run(httpAddr, httpsAddr string) error {
-}
-
-// REST Endpoint Handlers -----------------------
-
-func (s *Server) authenticate(w http.ResponseWriter, r *http.Request) {
-	authInfo := struct {
-		Username string `json:"username"`
-		Password string `json:"password"`
-	}{}
-
-	if err := json.NewDecoder(r.Body).Decode(&authInfo); err != nil {
-		http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
-		return
-	}
+func (s *Server) newRouter() *mux.Router {
+	router := mux.NewRouter()
 
-	user, err := s.store.User.Get(authInfo.Username)
-	if err != nil {
-		if err == store.ErrNoResults {
-			http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
-		} else {
-			s.logger.LogRequestJSON(r, map[string]interface{}{"error": fmt.Sprintf("retrieving user: %v", err)})
-			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
-		}
-		return
-	}
+	router.HandleFunc("/authenticate", s.authenticateHandler)
+	router.Handle("/users", s.authHandler(http.HandlerFunc(s.createUserHandler)))
+	router.Handle("/users/{username}", s.authHandler(http.HandlerFunc(s.deleteUserHandler)))
 
-	if err := bcrypt.CompareHashAndPassword([]byte(user.HashedPassword), []byte(authInfo.Password)); err != nil {
-		if err == bcrypt.ErrMismatchedHashAndPassword {
-			http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
-		} else {
-			s.logger.LogRequestJSON(r, map[string]interface{}{"error": fmt.Sprintf("comparing password hashes: %v", err)})
-			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
-		}
-		return
-	}
+	router.HandleFunc("/events", s.eventsHandler).Methods("GET")
+	router.Handle("/events/{eventKey}", s.pollMatchMiddleware(http.HandlerFunc(s.eventHandler))).Methods("GET")
+	router.Handle("/events/{eventKey}/{matchKey}", s.pollMatchMiddleware(http.HandlerFunc(s.matchHandler))).Methods("GET")
 
-	// from this point on the user has been successfully authenticated, just give them a token!
+	router.Handle("/reports/{eventKey}/{matchKey}", s.authHandler(http.HandlerFunc(s.reportHandler))).Methods("PUT")
 
-	ss, err := s.GenerateJWT(user.Username)
-	if err != nil {
-		s.logger.LogRequestJSON(r, map[string]interface{}{"error": fmt.Sprintf("generating jwt token: %v", err)})
-		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
-		return
-	}
+	router.HandleFunc("/analysis/{eventKey}", s.eventAnalysisHandler).Methods("GET")
+	router.HandleFunc("/analysis/{eventKey}/{team}", s.teamAnalysisHandler).Methods("GET")
+	router.HandleFunc("/analysis/{eventKey}/{matchKey}/{color}", s.allianceAnalysisHandler).Methods("GET")
 
-	respond.JSON(w, map[string]string{"jwt": ss})
+	return router
 }
 
-func (s *Server) getUser(w http.ResponseWriter, r *http.Request) {
-	vars := mux.Vars(r)
-
-	username := vars["username"]
+const tbaURL = "http://www.thebluealliance.com/api/v3"
```

Comment:
Would this be better as an argument to server? I guess we're not going to change versions of the api anytime soon, so it probably doesn't matter.

Reply:
If we changed API versions it would (probably) break our code, so imo it should be a constant.

Reply:
sg, that makes sense